qiskit/synthesis/evolution/product_formula.py

```diff
@@ -177,6 +186,79 @@ def evolve_pauli(
         _multi_qubit_evolution(output, pauli, time, cx_structure, wrap)
 
 
+def reorder_paulis(
+    operators: SparsePauliOp | list[SparsePauliOp],
+) -> SparsePauliOp | list[SparsePauliOp]:
+    r"""
+    Creates an equivalent operator by reordering terms in order to yield a
+    shallower circuit after evolution synthesis. The original operator remains
+    unchanged.
+
+    This method works in three steps. First, a graph is constructed, where the
+    nodes are the terms of the operator and where two nodes are connected if
+    their term acts on the same qubit (for example, the terms $IXX$ and $IYI$
+    would be connected, but not $IXX$ and $YII$). Then, the graph is colored.
+    Two terms with the same color thus do not act on the same qubit, and in
+    particular, their evolution subcircuits can be run in parallel in the
+    greater evolution circuit of ``operator``. Finally, a new
+    :class:`~qiskit.quantum_info.SparsePauliOp` is created where terms of the
+    same color are grouped together.
+
+    In trivial cases, i.e. when either
+    - the input is a :class:`~qiskit.quantum_info.SparsePauliOp` with
+      less than two Pauli terms, or
+    - the input is a list containing a single
+      :class:`~qiskit.quantum_info.SparsePauliOp` which has less than two
+      Pauli terms,
+    this method does nothing. In non-trivial cases, this method always returns a
+    :class:`~qiskit.quantum_info.SparsePauliOp` (rather than a list of such),
+    even if ``operators`` is a list.
+
+    This method is deterministic and invariant under permutation of the Pauli term in ``operators``.
+
+    Args:
+        operators: The operator or list of operators whose terms to reorder.
+    """
+
+    def _term_sort_key(term: tuple[str, complex]) -> Any:
+        return (term[0], term[1].real, term[1].imag)
+
+    # Do nothing in trivial cases
+    if isinstance(operators, SparsePauliOp) and len(operators) <= 1:
+        return operators
+    if isinstance(operators, list) and not (
+        # operators is a list of > 1 SparsePauliOp
+        len(operators) > 1
+        # operators has a single SparsePauliOp, which has > 1 terms
+        or (len(operators) == 1 and len(operators[0]) > 1)
+    ):
+        return operators
+    if not isinstance(operators, list):
+        operators = [operators]
+
+    terms = sum((o.to_list() for o in operators), [])
+    terms = sorted(terms, key=_term_sort_key)
+    graph = rx.PyGraph()
+    graph.add_nodes_from(terms)
+    indexed_nodes = list(enumerate(graph.nodes()))
+    for (idx1, term1), (idx2, term2) in combinations(indexed_nodes, 2):
+        # Add an edge between term1 and term2 if they touch the same qubit
+        for a, b in zip(term1[0], term2[0]):
+            if not (a == "I" or b == "I"):
+                graph.add_edge(idx1, idx2, None)
+                break
```

Comment:
I would think there is a better way to do this based on the `PauliList` boolean array saved inside the `SparsePauliOp` instances. A simple way is the following:
```python
x_or_z = op.paulis.x | op.paulis.z
overlaps = x_or_z @ x_or_z.T
```
This yields a matrix indicating on entry `(i, j)` whether the Pauli term at index `i` has overlap with that at index `j`.

This computes the full matrix but you are really only interested in the upper triangular indices. So there may be an even more efficient way to do this...

Reply:
Actually, this might also help with the algorithm becoming deterministic without requiring the terms to be sorted first :thinking:

Reply:
This is now obsolete as we deal with sparse pauli labels 🙂